File: SaffronEngine/Collection/ViewportPane.cs
using System;
using System.Numerics;
using System.Text;
using ImGuiNET;
using Renderer;
using SaffronEngine.Common;
using SaffronEngine.Rendering;
using Serilog;

namespace SaffronEngine.Collection
{
    public class ViewportPane
    {
        private readonly string _viewportTitle;
        private uint _dockId;
        public SceneRenderer.Target Target { get; set; }
        private IntPtr _viewportGuiTextureHandle;

        private Vector2 _topLeft;
        private Vector2 _bottomRight;

        private readonly uint _inactiveBorderColor = BitConverter.ToUInt32(new byte[] {255, 140, 0, 80});
        private readonly uint _activeBorderColor = BitConverter.ToUInt32(new byte[] {255, 140, 0, 180});

        public ViewportPane(string viewportTitle, SceneRenderer.Target target)
        {
            this._viewportTitle = viewportTitle;

            _viewportGuiTextureHandle = Gui.AddTexture(target.FrameBuffer.GetTexture());
            Target = target;
            Target.Resized += (sender, args) =>
            {
                Gui.RemoveTexture(_viewportGuiTextureHandle);
                Gui.AddTexture(Target.FrameBuffer.GetTexture());
            };

            TopLeft = Vector2.Zero;
            BottomRight = Vector2.Zero;
        }

        public void OnGuiRender()
        {
            var tl = TopLeft;
            var br = BottomRight;

            ImGui.PushStyleVar(ImGuiStyleVar.WindowPadding, Vector2.Zero);

            const int uuid = 0;

            var builder = new StringBuilder();
            builder.Append(_viewportTitle).Append("##").Append(uuid);

            ImGui.Begin(builder.ToString(), ImGuiWindowFlags.NoFocusOnAppearing);

            _dockId = ImGui.GetWindowDockID();

            Hovered = ImGui.IsWindowHovered();
            Focused = ImGui.IsWindowFocused();

            var viewportOffset = ImGui.GetCursorPos(); // includes tab bar
            var minBound = ImGui.GetWindowPos();
            minBound.X += viewportOffset.X;
            minBound.Y += viewportOffset.Y;

            var windowSize = ImGui.GetWindowSize();
            var maxBound = new Vector2(minBound.X + windowSize.X - viewportOffset.X,
                minBound.Y + windowSize.Y - viewportOffset.Y);
            _topLeft.X = minBound.X;
            _topLeft.Y = minBound.Y;
            _bottomRight.X = maxBound.X;
            _bottomRight.Y = maxBound.Y;

            var vpSize = ViewportSize;
            var fbTexture = Target.FrameBuffer.GetTexture();
            var imageRendererId = fbTexture.GetHashCode();

            ImGui.Image((IntPtr) imageRendererId, new Vector2(vpSize.X, vpSize.Y));

            ImGui.GetWindowDrawList().AddRect(new Vector2(TopLeft.X, tl.Y), new Vector2(br.X, br.Y),
                Focused ? _activeBorderColor : _inactiveBorderColor, 0.0f, ImDrawCornerFlags.All, 4);

            ImGui.End();
            ImGui.PopStyleVar();


            if (!(Math.Abs(ViewportSize.X - Target.Width) > 0.5f) &&
                !(Math.Abs(ViewportSize.Y - Target.Height) > 0.5f)) return;

            Target.Width = (uint) ViewportSize.X;
            Target.Height = (uint) ViewportSize.Y;

            Resized?.Invoke(this,
                new SizeEventArgs(new SizeEvent {Width = (uint) ViewportSize.X, Height = (uint) ViewportSize.Y}));
        }

        public Vector2 TopLeft
        {
            get => _topLeft;
            private set => _topLeft = value;
        }

        public Vector2 BottomRight
        {
            get => _bottomRight;
            private set => _bottomRight = value;
        }

        public bool Hovered { get; private set; }
        public bool Focused { get; private set; }

        public Vector2 ViewportSize => BottomRight - TopLeft;

        public event EventHandler<SizeEventArgs> Resized = null;
    }
}

File: Engine/Source/Saffron/Editor/ViewportPane.cpp
#include "SaffronPCH.h"

#include "Saffron/Editor/ViewportPane.h"
#include "Saffron/Gui/Gui.h"
#include "Saffron/Input/Mouse.h"
#include "Saffron/Rendering/SceneRenderer.h"

namespace Se
{
ViewportPane::ViewportPane(String windowTitle) :
	_windowTitle(Move(windowTitle)),
	//_fallbackTexture(Texture2D::Create("Resources/Assets/Editor/FallbackViewportPaneTexture.png")),
	_topLeft(0.0f, 0.0f),
	_bottomRight(100.0f, 100.0f),
	_hovered(false),
	_focused(false)
{
}

void ViewportPane::OnGuiRender(bool* open, UUID uuid)
{
	const auto& tl = GetTopLeft();
	const auto& br = GetBottomRight();

	ImGui::PushStyleVar(ImGuiStyleVar_WindowPadding, ImVec2(0, 0));

	OutputStringStream oss;
	oss << _windowTitle << "##" << uuid;

	ImGui::Begin(oss.str().c_str(), open, ImGuiWindowFlags_NoFocusOnAppearing);

	_dockID = ImGui::GetWindowDockID();

	if (ImGui::IsWindowDocked())
	{
		auto* wnd = ImGui::FindWindowByName(_windowTitle.c_str());
		if (wnd)
		{
			ImGuiDockNode* node = wnd->DockNode;
			if (node && !node->IsHiddenTabBar())
			{
				node->WantHiddenTabBarToggle = true;
			}
		}
	}

	_hovered = ImGui::IsWindowHovered();
	_focused = ImGui::IsWindowFocused();

	const auto viewportOffset = ImGui::GetCursorPos(); // includes tab bar
	ImVec2 minBound = ImGui::GetWindowPos();
	minBound.x += viewportOffset.x;
	minBound.y += viewportOffset.y;

	const auto windowSize = ImGui::GetWindowSize();
	const ImVec2 maxBound = {
		minBound.x + windowSize.x - viewportOffset.x, minBound.y + windowSize.y - viewportOffset.y
	};
	_topLeft = {minBound.x, minBound.y};
	_bottomRight = {maxBound.x, maxBound.y};

	const auto viewportSize = GetViewportSize();
	const auto imageRendererID = SceneRenderer::GetFinalColorBufferRendererID();
	ImGui::Image(reinterpret_cast<void*>(imageRendererID), {viewportSize.x, viewportSize.y}, {0, 1}, {1, 0});
	ImGui::GetWindowDrawList()->AddRect(ImVec2(_topLeft.x, tl.y), ImVec2(br.x, br.y),
	                                    _focused ? IM_COL32(255, 140, 0, 180) : IM_COL32(255, 140, 0, 80), 0.0f,
	                                    ImDrawCornerFlags_All, 4);

	FinishedRender.Invoke();

	ImGui::End();
	ImGui::PopStyleVar();

	SceneRenderer::SetViewportSize(static_cast<Uint32>(viewportSize.x), static_cast<Uint32>(viewportSize.y));
}

bool ViewportPane::InViewport(Vector2f positionNDC) const
{
	positionNDC.x -= _topLeft.x;
	positionNDC.y -= _topLeft.y;
	return positionNDC.x < _bottomRight.x && positionNDC.y < _bottomRight.y;
}

Vector2f ViewportPane::GetMousePosition() const
{
	Vector2f position = Mouse::GetPositionNDC();
	position.x -= _topLeft.x;
	position.y -= _topLeft.y;
	const auto viewportWidth = _bottomRight.x - _topLeft.x;
	const auto viewportHeight = _bottomRight.y - _topLeft.y;

	return {position.x / viewportWidth * 2.0f - 1.0f, (position.y / viewportHeight * 2.0f - 1.0f) * -1.0f};
}

Vector2f ViewportPane::GetViewportSize() const
{
	return GetBottomRight() - GetTopLeft();
}
}
